File: commons/python/aoc/answer.py
import time
from functools import wraps
from typing import Callable


def timer(solution: Callable[[], None]) -> Callable[[], None]:
    @wraps(solution)
    def wrapper() -> None:
        start = time.time_ns()
        solution()
        end = time.time_ns()
        print(f"Runtime (ns): {end - start}")

    return wrapper


def part1[T](expected: T, actual: T) -> None:
    part(1, expected, actual)


def part2[T](expected: T, actual: T) -> None:
    part(2, expected, actual)


def part[T](part: int, expected: T, actual: T) -> None:
    if expected != actual:
        raise Exception(f"Part {part} incorrect, expected {expected} but got {actual}")
    print(f"Part {part}: {actual}")

File: commons/go/answer/answer.go
package answer

import (
	"fmt"
	"time"
)

func Timer(solution func()) {
	start := time.Now().UnixNano()
	solution()
	end := time.Now().UnixNano()
	fmt.Printf("Runtime (ns): %d\n", end-start)
}

func Part1[T comparable](expected, actual T) {
	part(1, expected, actual)
}

func Part2[T comparable](expected, actual T) {
	part(2, expected, actual)
}

func part[T comparable](part int, expected T, actual T) {
	if expected != actual {
		panic(fmt.Sprintf("Part %d incorrect, expected %v but got %v", part, expected, actual))
	}
	fmt.Printf("Part %d: %v\n", part, actual)
}
